Firestore Watch

blowmage opened the pull request:
This PR adds the ability to listen to a document reference or a collection reference/query for changes. A document or query snapshot will be yielded immediately, and each time the contents change.

You can use `DocumentReference#listen` to be notified of changes to a single document:

```ruby
require "google/cloud/firestore"

firestore = Google::Cloud::Firestore.new

# Get a document reference
nyc_ref = firestore.doc "cities/NYC"

listener = nyc_ref.listen do |snapshot|
  puts "The population of #{snapshot[:name]} "
  puts "is #{snapshot[:population]}."
end

# When ready, stop the listen operation and close the stream.
listener.stop
```

You can use `Firestore::Query#listen` to be notified of changes to any document contained in the query:

```ruby
require "google/cloud/firestore"

firestore = Google::Cloud::Firestore.new

# Create a query
query = firestore.col(:cities).order(:population, :desc)

listener = query.listen do |snapshot|
  puts "The query snapshot has #{snapshot.docs.count} documents "
  puts "and has #{snapshot.changes.count} changes."
end

# When ready, stop the listen operation and close the stream.
listener.stop
```

This is accomplished by the following:

* Add `DocumentReference#listen` and `Query#listen` methods.
* Add `QuerySnapshot` and `DocumentChange` classes.
* Add `DocumentListener` and `QueryListener` classes.
* `DocumentReference#listen` yields a `DocumentSnapshot` object.
* `Query#listen` yields a `QuerySnapshot` object.
* Both `listen` methods return a listener object (`DocumentListener` or `QueryListener`) that users can use to control the watch stream.

blowmage commented:
@schmidt-sebastian I can't add you as a reviewer. Can your look over these changes and submit a review?

blowmage commented on google-cloud-firestore/lib/google/cloud/firestore/document_listener.rb:
```diff
@@ -0,0 +1,284 @@
+            break if synchronize { @stopped }
+
+            begin
+              # Cannot syncronize the enumerator, causes deadlock
```
Yes, it could. Originally it was shared between the DocumentListener and QueryListener, but it was unfactored because of the churn during development. It can be extracted in a future refactoring, but I would like to make sure it is stable first. Duplication is easier to deal with than a bad abstraction.

blowmage commented on google-cloud-firestore/lib/google/cloud/firestore/document_listener.rb:
```diff
@@ -0,0 +1,284 @@
+                Google::Firestore::V1beta1::Target::DocumentsTarget.new(
+                  documents: [@doc_ref.path]
+                ),
+              target_id: 0x42
```
Yep. Good catch.

blowmage replied:
Fixed.

blowmage commented on google-cloud-firestore/lib/google/cloud/firestore/query_listener.rb:
```diff
@@ -0,0 +1,473 @@
+            # This is because the API returns sorted query results.
+            # This also provides a massive performance improvement because
+            # sorting an already sorted red-black tree is a worst case scenario.
+            return 1 if @initial_load
```
Good to know. I thought I read somewhere that we should trust the order of the documents on the initial load. This makes inserting large numbers of already-sorted documents expensive, as ordered insert is a worst case scenario for the red/black tree library we are using. The benchmark for inserting 25,000 documents is scary.

blowmage replied:
Unfortunately this was removed. I might have to perform a performance pass in a follow-up PR.

blowmage commented on google-cloud-firestore/lib/google/cloud/firestore/query_listener.rb:
```diff
@@ -0,0 +1,473 @@
+            return 3 if value.is_a? String
+            return 4 if value.is_a? StringIO
+            return 5 if value.is_a? Array
+            return 6 if value.is_a? Hash
```
I don't think GeoPoints are implemented currently. We also don't allow inserting a vanilla Ruby Object in a document's data. We only support the specified types, same as Datastore and Spanner.

blowmage replied:
Oh, we use a special Hash to map to/from GeoPoint. I'll make sure that is supported properly. Good catch.

blowmage replied:
We have identified and fixed a bug with the GeoPoint mapping.

blowmage commented on google-cloud-firestore/lib/google/cloud/firestore/query_listener.rb:
```diff
@@ -0,0 +1,473 @@
+          end
+
+          def field_comparison value
+            [field_type(value), field_value(value)]
```
Good catch, I'll add the check for NaN. Ruby comparisons already match Firestore semantics for non-NaN Float values.

```
$ irb
irb(main):001:0> -0.0 <=> +0.0
=> 0
irb(main):002:0> 0 <=> Float::INFINITY
=> -1
irb(main):003:0> 0 <=> -Float::INFINITY
=> 1
```

blowmage replied:
This is updated and we have tests to show this is working properly.

blowmage commented on google-cloud-firestore/lib/google/cloud/firestore/query_listener.rb:
```diff
@@ -0,0 +1,473 @@
+            return 1 if value == false
+            return 1 if value == true
+            return 2 if value.is_a? Numeric
+            return 3 if value.is_a? String
```
I don't follow. What is a Date?

schmidt-sebastian replied:
Type Order '3' is a Firestore Timestamp (aka Date).

This order needs to match the backend exactly. It is crucial to match the backend ordering, as not doing so might cause you to miss documents as you paginate through the results (via startAt() and its friends).

blowmage replied:
We have added the missing support for NaN, Timestamp, and GeoPoint. Including new unit tests and acceptance tests for these values.

blowmage commented on google-cloud-firestore/lib/google/cloud/firestore.rb:
```diff
@@ -390,6 +390,52 @@ module Cloud
+    # ### Listening for changes
+    #
+    # You can listen to a document reference or a collection reference/query for
+    # changes. A document or query snapshot will be yielded immediately, and
```
I think what we are trying to say here is that when you start listening, you get a snapshot of things as they are right now, and then you will get snapshots of changes as they happen.

For example, when you start listening to a document that isn't being changed, you will always get a snapshot of the document, but you won't get second snapshot because there are no changes being made.

quartzmo replied:
@schmidt-sebastian We adapted that documentation from [Get realtime updates with Cloud Firestore](https://firebase.google.com/docs/firestore/query-data/listen):

> You can listen to a document with the onSnapshot() method. An initial call using the callback you provide creates a document snapshot immediately with the current contents of the single document. Then, each time the contents change, another call updates the document snapshot.

If it is incorrect, let us know and we will remove it.

blowmage replied:
How about replacing

> A document or query snapshot will be yielded immediately, and

with:

> The current document snapshot or query results snapshot will be yielded first, and

schmidt-sebastian replied:
Sounds good.

blowmage commented on google-cloud-firestore/lib/google/cloud/firestore/document_change.rb:
```diff
@@ -0,0 +1,123 @@
+      #
+      #   listener = query.listen do |snapshot|
+      #     puts "The query snapshot has #{snapshot.docs.count} documents "
+      #     puts "and has #{snapshot.changes.count} changes."
```
Yeah, we discussed how detailed to get in the code examples, and we ultimately decided showing that changes were there was good enough.

The only lib that has any code examples for DocumentChange looks to be [node.js](https://cloud.google.com/nodejs/docs/reference/firestore/0.15.x/DocumentChange). (/cc @quartzmo)

blowmage commented on google-cloud-firestore/lib/google/cloud/firestore/query_listener.rb:
```diff
@@ -0,0 +1,473 @@
+          # Has the loop broken but we aren't stopped?
+          # Could be GRPC has thrown an internal error, so restart.
+          synchronize { raise "restart thread" unless @stopped }
+        rescue GRPC::DeadlineExceeded, GRPC::Unavailable, GRPC::Cancelled,
```
@schmidt-sebastian I'm working on this right now. Do we need exponential backoff on all these errors? Or just UNAVAILABLE?

blowmage replied:
I pushed an update that implements incremental backoff for UNAVAILABLE only. I believe this is the only request not addressed so far. Thoughts?

blowmage replied:
What errors are considered "reconnect"?

schmidt-sebastian replied:
Please take a look here: https://github.com/googleapis/nodejs-firestore/blob/master/src/watch.js#L295

blowmage commented on google-cloud-firestore/lib/google/cloud/firestore/query_listener.rb:
```diff
@@ -0,0 +1,473 @@
+            new_order = Hash[docs.map(&:path).each_with_index.to_a] # O(n)
+            changes = calc_changes \
+              @to_add, @to_delete, @old_order, new_order, read_time
+
```
This has been completely rewritten in order to support the conformance tests that were just added this week. We no longer track changes, we calculate changes from the last snapshot.

blowmage commented on google-cloud-firestore/lib/google/cloud/firestore/query_listener.rb:
```diff
@@ -0,0 +1,473 @@
+            path_comparison(a_path) <=> path_comparison(b_path)
+          end
+
+          def field_comparison value
```
I'm going to keep it here for now. We can move these methods to a new location once we have a need to reuse them.

blowmage commented on google-cloud-firestore/test/helper.rb:
```diff
@@ -42,6 +42,22 @@ def == other
+  end
+end
+
 class MockFirestore < Minitest::Spec
```
This has been added. See `google-cloud-firestote/test/google/cloud/firestore/watch/order_test.rb`.

blowmage commented on google-cloud-firestore/lib/google/cloud/firestore/query_listener.rb:
```diff
@@ -0,0 +1,566 @@
+          # Has the loop broken but we aren't stopped?
+          # Could be GRPC has thrown an internal error, so restart.
+          raise RestartStream
+        rescue GRPC::DeadlineExceeded, GRPC::Unavailable, GRPC::Cancelled,
```
@schmidt-sebastian Can you verify the errors to retry with incremental backoff list here is correct? This is the same list that Spanner is using to restart streams with.

schmidt-sebastian replied:
This is not the list we use in the other clients, but I am generally ok with keeping this consistent with Spanner.  The one exception to this is `UNAUTHENTICATED`. Please verify that you can re-open the stream when the authentication expires. This will happen after 60 minutes.

blowmage replied:
I’ll add UNAUTHENTICATED.

blowmage replied:
This has been added.

schmidt-sebastian commented on google-cloud-firestore/lib/google/cloud/firestore/query_listener.rb:
```diff
@@ -0,0 +1,566 @@
+          end
+
+          def current_docs
+            @tree.keys
```
Does this create a copy? You need to ensure that changes to the `inventory` don't affect QuerySnapshots that have already been raised.

blowmage replied:
No, the red-black tree library we are using does not create a copy. Instead, we are storing the ordered document paths and update time in a hash. We have an old hash from the previous callback, and a new hash for the current callback. We calculate which docs have been added, removed, and updated (including their indexes) using these hashes.

Changing a document snapshot in inventory removes the old document snapshot object from the tree and then adds the new document snapshot object to the tree. A query snapshot that has been passed to the callback will not be affected, because the document snapshot objects are immutable.

Target changes are pulled from the stream in order, and are handled sequentially. The query snapshot will be sent to the callback (raised) before another target change is handled.

blowmage replied:
Yes, each query snapshot and document snapshot are considered immutable. Previous snapshots and future snapshots won't change the current snapshot contents.

schmidt-sebastian commented on google-cloud-firestore/lib/google/cloud/firestore/query_listener.rb:
```diff
@@ -0,0 +1,566 @@
+                # that target).
+
+                @inventory.delete response.document_remove.document
+              end
```
This is missing the Existing Filter handling. See https://github.com/googleapis/nodejs-firestore/blob/master/src/watch.js#L762

blowmage replied:
I don't see any documentation about this. What are we supposed to do for this? It looks like the node implementation does a full reset when the count is off. But the proto documentation says:

> If different from the count of documents in the client that match, the client must manually determine which documents no longer match the target.

How do we determine if the documents match the target or not?

blowmage replied:
Okay, thanks. I will add this.

Bloom filters are... fun. Sorta.

blowmage replied:
This has been added. Please take a look.

You will be happy to know that calculating the inventory count was the functionality that ultimately forced my hand to extract the listen functionality to a common object. `DocumentListener` and `QueryListener` still exist, because their callback mechanics and documentation are different, but they now use a common `Watch::Listener` object to do all the heavy lifting.

schmidt-sebastian commented on google-cloud-firestore/lib/google/cloud/firestore/query_listener.rb:
```diff
@@ -0,0 +1,566 @@
+            return 1 if value == false
+            return 1 if value == true
+            return 2 if value.respond_to?(:nan?) && value.nan?
+            return 3 if value.is_a? Numeric
```
Hm, this is a pretty clever handling of the NaN case, but I am still a little bit opposed. We don't have a sane way to enforce ordering across all SDKs and the backend if the code doesn't match.

blowmage replied:
I assume this comment was intended for line 474, and not 475. Can you confirm that your concern is separating NaN values from every other numeric value?

If so, we are doing this because of the way Ruby's comparison works for NaN values. Like `nil`, they don't compare to each other.

```ruby
num1 = Float::NAN
num2 = Float::NAN
num1 == num2 #=> false
```

Worse, comparing a NaN value will return a `nil`, which is considered and error case:

```ruby
num1 <=> num2 #=> nil
```

So we separate them so all NaN values are ordered before all other numbers (or after if the order is descending), and we don't actually compare the NaN values, we compare 0 instead. I believe we are enforcing order correctly, and have tests demonstrating this.

schmidt-sebastian replied:
Yes, that was my concern (line 474).

Can you add a comment here that is divergence is by design and that it should not be copied to other clients? Something like

```
The backend ordering semantics treats NaN and Numbers as the same type, and then internally orders NaNs before Numbers. In Ruby, we can achieve the same by separating the types and thus, we intentionally diverge from type order that is used in the backend and in the other SDKs.
```

blowmage replied:
Added.

blowmage commented on google-cloud-firestore/lib/google/cloud/firestore/client.rb:
```diff
@@ -382,31 +382,37 @@ def batch
           rescue Google::Cloud::InvalidArgumentError => err
-            # Return if a previous call has succeeded
-            return nil if retries > 0
+            # Return if a previous call was retried but ultimately succeeded
```
Can we move this discussion to a separate issue? We haven't made any behavioral changes to transactions in the PR, and I'd like to keep this as focused as we can.

Our new policy is to squash all commits when the PR is merged, so it makes it much more difficult to track unrelated changes.

blowmage commented on google-cloud-firestore/lib/google/cloud/firestore/document_listener.rb:
```diff
@@ -0,0 +1,305 @@
+
+                  unless @inventory.empty?
+                    synchronize do
+                      send_callback get_latest_doc_snp(@read_time)
```
Agreed, this was missing and has been added.

blowmage commented on google-cloud-firestore/lib/google/cloud/firestore/document_listener.rb:
```diff
@@ -0,0 +1,305 @@
+          raise err if @backoff[:current] >= @backoff[:max]
+
+          # Sleep with incremental backoff before restarting
+          sleep @backoff[:delay]
```
I refactored this a bit, please take a look again. Hopefully it is more understandable now.

blowmage commented on google-cloud-firestore/lib/google/cloud/firestore/query_listener.rb:
```diff
@@ -0,0 +1,566 @@
+                    )
+                  )
+
+                  if @inventory.current? && @inventory.changes?
```
As I mentioned before, I separated the implementations because it was too difficult to make changes to. I consider this PR to still be in the red phase of red/green/refactor. Once the PR is approved then we can try refactoring the implementation using the test coverage we have built. But for now I am leaving these implementations separate.

blowmage commented on google-cloud-firestore/lib/google/cloud/firestore/query_listener.rb:
```diff
@@ -0,0 +1,566 @@
+                  # returned even if the target was previously indicated to be
+                  # +CURRENT+.
+
+                  raise RestartStream # Raise to restart the stream
```
Ah, this was new to me. I've added this.

blowmage commented:
@schmidt-sebastian I've been working on performance tuning the Watch implementation. I had to rebase my performance changes on top of the changes you requested, and I decided to add them here to this PR as it seems we've addressed all of your concerns. The performance tuning mostly consists of memoizing document comparison values, which seems to increase performance by ~25%.

The red-black tree we are using is self-balancing, and that means adding a value to the tree is O(log(n)). We don't really see this being a problem until a query snapshot has ~2000 documents in it. That is when the initial insertion of all the documents into the tree data structure starts to becomes slower than a simple quick sort. After that, incremental adding a document is much faster with the tree, but if enough documents are added between snapshots then it becomes faster to use quick sort again. That number seems to be ~500 on my dev computer. Deleting a document is also more CPU intensive than removing from a Ruby array, but that is to be expected.

I don't know how these thresholds compare to other implementations, but I'm fairly content with the improvements made.

blowmage commented:
@schmidt-sebastian Can you review these changes?

schmidt-sebastian commented on google-cloud-firestore/lib/google/cloud/firestore/document_snapshot.rb:
```diff
@@ -285,11 +308,47 @@ def missing?
+        # @private
+        def eql? other
+          return nil unless other.is_a? DocumentSnapshot
+          grpc.eql? other.grpc
```
Is `grpc` the backing Protobuf? We have decided to not include the metadata (read time, update time, create time) in the equals function since it makes the API practically unusable. No Documents would be the same unless their pointer addresses are the same.

Feel free to address in a follow up if you think this is an actual issue.

blowmage replied:
Yes, `grpc` is the protobuf object. 

This will affect not only equals, but the hash value as well. I’ll adjust these methods to ignore these date values.

Should empty snapshots be ordered before or after snapshots with data? Assuming the resource names are the same?

schmidt-sebastian replied:
That should never happen (you should never have the same document with two different sets of data in the same data structure). If you do want to address this, then feel free to pick an ordering that suits your liking.

blowmage replied:
These methods are private and here to make things like lookup in a Hash more efficient. They are not used by the Watch ordering. I've updated them to not consider the date values.

schmidt-sebastian commented on google-cloud-firestore/lib/google/cloud/firestore/watch/listener.rb:
```diff
@@ -0,0 +1,297 @@
+            init_listen_req = @init_listen_req.dup.tap do |req|
+              req.add_target.resume_token = String(@inventory.resume_token)
+              req.add_target.read_time = \
+                Convert.time_to_timestamp(@inventory.read_time)
```
You should only one of the two - either `resume_token` or `read_time`. We only use the `resume_token` in the other clients.

blowmage replied:
Okay. I’ll remove `read_time` in this cal.

blowmage replied:
This has been removed.

schmidt-sebastian commented on google-cloud-firestore/lib/google/cloud/firestore/watch/listener.rb:
```diff
@@ -0,0 +1,297 @@
+                    # targets will be returned in subsequent changes.
+                    #
+                    # After the initial state is complete, +CURRENT+ will be
+                    # returned even if the target was previously indicated to be
```
s/returned/reset/?

blowmage replied:
This is the documentation copied from the protobuf. It is copied here for reference only.

blowmage replied:
IOW I don’t know if reset or returned is correct.

schmidt-sebastian commented on google-cloud-firestore/lib/google/cloud/firestore/watch/listener.rb:
```diff
@@ -0,0 +1,297 @@
+                  # Returned when documents may have been removed from the given
+                  # target, but the exact documents are unknown.
+
+                  if response.filter.count != @inventory.size
```
This needs to be the current size with all pending changes applied.

blowmage replied:
So apply all pending changes, compare the count, then move the applied pending changes back to pending?

schmidt-sebastian replied:
Yes, that would work.

blowmage replied:
Okay, this has been updated to apply the pending changes when calculating the count.

schmidt-sebastian replied:
Looks good.
